pkg/controller/replicaset/replica_set.go

```diff
@@ -819,7 +819,9 @@ func getPodsRankedByRelatedPodsOnSameNode(podsToRank, relatedPods []*v1.Pod) con
 	}
 	ranks := make([]int, len(podsToRank))
 	for i, pod := range podsToRank {
-		ranks[i] = podsOnNode[pod.Spec.NodeName]
+		if ranks[i] = podsOnNode[pod.Spec.NodeName]; ranks[i] > 0 {
+			podsOnNode[pod.Spec.NodeName]--
```

Comment:
This is removing the usefulness of other criteria, like the time the pods have been running.
Can you think of an alternative to not lose that?

Reply:
cc @damemi who has been looking into this.

Reply:
Yeah, it would probably be better to try to tie this into the `sort` function that is already called on these podsWithRanks, as done in `getPodsToDelete()`. See https://github.com/kubernetes/kubernetes/blob/d8aad7944cc71314913fc0b43d6731e2c36f4928/pkg/controller/controller_utils.go#L820-L894

Reply:
One alternative I can think of is to do the sorting without ranks first, and then run the loop to add the ranks and finally do a second stable sort.

Reply:
Or... we can simply remove the ranks. Now that we have randomized removals https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/2185-random-pod-select-on-replicaset-downscale, the nodes with higher number of pods are more likely to get pods removed :)

Reply:
It's only random among similar time scales though, so is it guaranteed that nodes with more pods are equally likely as nodes with fewer pods? For example you could have a long-running node with many replicas and a fresh node with only 1 or 2, that newer node will still be preferred by age right?

Reply:
Correct. And that might be fine. If it's a newer node, perhaps it was recently scaled up. If the load is now lower, you might want to get rid of that new node to reduce your cost. It's all heuristics.
Ultimately, users can now use the [pod deletion cost](https://kubernetes.io/docs/concepts/workloads/controllers/replicaset/#pod-deletion-cost) feature to fine tune the behavior. But the bucketed randomness seems good enough to me for most use cases.

Reply:
@alculquicondor @damemi Thanks you guys! I understand the random KEP, but what do you think of this PR? Currently this unexpected deletion rank is bothering us, shouldn't we fix it?
Sure users can set the pod-deletion-cost to customize deletion priority, but I think it is not the reason that replicaset controller could keep this incorrect logic.

BTW, speak of ReplicaSet with an imbalanced state across failure domains, have you considered that ReplicaSet downscale should also take Pod Topology Spread Constraints as its rank too?

Reply:
> have you considered that ReplicaSet downscale should also take Pod Topology Spread Constraints as its rank too?

That was actually the original proposal for the random downscale KEP, but due to the complexity that level of consideration was put on hold, see https://github.com/kubernetes/kubernetes/issues/96748

Reply:
@FillZpp the problem is that your PR as currently written is essentially disabling the random downscale.

So my proposal is this: when random downscale is enabled, do not put ranks. That should address (probabilistically) what you want to fix.

Reply:
@alculquicondor I have read through the [random selection kep](https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/2185-random-pod-select-on-replicaset-downscale). It is a good design, but I think it should aim to replace the directly sorting by ready or creation timestamp, rather than the node rank.

The reasonable sequence might be:
1. node unassigned < assigned
2. PodPending < PodUnknown < PodRunning
3. not ready < ready
4. lower pod-deletion cost < higher pod-deletion-cost (btw, the [code comment here](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/controller_utils.go#L836) is wrong)
5. spread pods on the same node by rank
6. use randomized algorithm instead of directly sorting by ready or creation timestamp

Reply:
how do you plan to achieve 5?

Reply:
> how do you plan to achieve 5?

Emm... You have confused me. Isn't it the logic I want to fix that helps spread pods on the same node?

Reply:
This PR? It does, but the way it assigns the ranks causes (6) to never be checked.

For all the pods that are in a node, we want them to be sorted according to bucketed random selection.

Do you agree that simply removing (5) solves the problem probabilistically?

Reply:
@alculquicondor But according to the code, `LogarithmicScaleDown` only take effect on the timestamp sort. (5) is still there, and I think it should not be removed directly.
https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/controller_utils.go#L849

```
        // 5. Doubled up < not doubled up
	// If one of the two pods is on the same node as one or more additional
	// ready pods that belong to the same replicaset, whichever pod has more
	// colocated ready pods is less
	if s.Rank[i] != s.Rank[j] {
		return s.Rank[i] > s.Rank[j]
	}

	// 6. Been ready for empty time < less time < more time
	// If both pods are ready, the latest ready one is smaller
	if podutil.IsPodReady(s.Pods[i]) && podutil.IsPodReady(s.Pods[j]) {
		// features.LogarithmicScaleDown ...
	}
	// 7. Pods with containers with higher restart counts < lower restart counts
	if maxContainerRestarts(s.Pods[i]) != maxContainerRestarts(s.Pods[j]) {
		return maxContainerRestarts(s.Pods[i]) > maxContainerRestarts(s.Pods[j])
	}
	// 8. Empty creation time pods < newer pods < older pods
	if !s.Pods[i].CreationTimestamp.Equal(&s.Pods[j].CreationTimestamp) {
		// features.LogarithmicScaleDown ...
	}
```

Reply:
yes, but such comparison only happens if the ranks are different. The way you are assigning ranks in this PR causes the pods within a node not to be compared by time at all.

What I'm suggesting is that, if you remove the rank completely, then all pods will be compared simply by bucketed time and otherwise randomized. Then, a node that has more pods has more changes to have its pods removed, obtaining the effect that you want to achieve. But it's just probabilistic. I think that's good enough.

Reply:
I think these two logic don't conflict. With this PR, pods with the same rank can still sort by randomized.

For example, there are 5 nodes and each of them has 2 pods on it. So, the rank of 5 pods on the different 5 nodes will be 1 and the other 5 pods will be 0. The random algorithm will sort the former 5 pods by randomized and the latter 5 pods by randomized.

![image](https://user-images.githubusercontent.com/7541475/123199111-d4a0c180-d4e0-11eb-9bf5-38b74a194862.png)

@alculquicondor @damemi  What do you think?

Reply:
But there is no comparison between a-0 and a-1. You are directly assigning a preference without checking the timestamp.

Reply:
> You are directly assigning a preference without checking the timestamp.

Yes, that might need to sort the pods on same node by timestamp and then give them the right ranks?

Reply:
Correct. I think it's only possible through two consecutive sort operations, as I proposed above. However, I think if we get rid of the rank entirely, we get the same result, probabilistically.

Reply:
WDYT?